DEPRECATIONS.md

```diff
@@ -0,0 +1,66 @@
+# Deprecations & Breaking Changes in KEDA
+
+This document describes how KEDA handles deprecations and breaking changes and should provide clarity to KEDA end-users on what to expect when upgrading to a new version of KEDA.
+
+If you are interested in the current open deprecations, we recommend checking our [GitHub Discussions](https://github.com/kedacore/keda/discussions/categories/deprecations) or the open [breaking changes](https://github.com/kedacore/keda/issues?q=is%3Aopen+label%3Abreaking-change+sort%3Aupdated-desc).
+
+## Our versioning strategy
+
+KEDA follows the deprecation strategy of Kubernetes and is allowed to make breaking changes in minor versions after a deprecation period.
+
+The KEDA runtime, which consists of the operator and metric server, will provide support on the deprecated features for 4 releases/1 year after the deprecation was announced. In the following release, the feature will be removed.
```

Comment:
```suggestion
The KEDA runtime, which consists of the operator and metric server, will provide support on the deprecated features for 4 releases/1 year after the deprecation was announced. In the following release, the feature will be removed.

In some scenarios, a graceful phase-out approach can be used by using feature flags. This allows end-users to keep using the deprecated functionality if they need to.

The phase-out with feature flags would be as follow:
1. A new feature flag is introduced at time of deprecation that enables the deprecated functionality to stay available. If end-users do not depend on it, they can already disable it to avoid taking new dependency on it.
2. In release N+2, the feature flag is turned off by default so that the deprecated functionality is no longer available as a scream test. However, if end-users still rely on it they can enable the functionality again.
3. N+4 releases removes support for the deprecated feature and it will no longer be available
```

Reply:
I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable then. We are already giving a year for the migration

Reply:
-1 we should try to keep things simple, we should not try to over engineer this. I'd be also for reducing the deprecation from 4 releases to 2 :)

Reply:
I'd reduce it too, but we decided 4 xD
My question is, does make sense a feature flag for removing features? Isn't the required effort quite similar to just upgrade/downgrade KEDA itself? I mean, changing the yaml to enable/disable the flag looks like just change the version.

Reply:
> I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable them (apart from the needed code, ofc). We are already giving a year for the migration

This is to get end-users off of it faster instead of only after 1 year.



> I'd reduce it too, but we decided 4 xD

Exactly, why did we decide to 4 if we change it again :)

> My question is, does make sense a feature flag for removing features? Isn't the required effort quite similar to just upgrade/downgrade KEDA itself? I mean, changing the yaml to enable/disable the flag looks like just change the version.

No because if we don't have flags we have to have full support for 4 versions long while with feature flags we could start "hiding" things so that less people use it by default.

Reply:
If nobody wants this then we can leave this out but this is exactly what we are considering at the moment :)

Reply:
> Exactly, why did we decide to 4 if we change it again :)

I don't want to change it, even I'd do it xD (We discussed this, and I respect our decision)

> No because if we don't have flags we have to have full support for 4 versions long while with feature flags we could start "hiding" things so that less people use it by default.

This is a user decision, if end-user decides to use a deprecated feature as default, it's totally their problem IMO, we cannot do anything there as they can enable the flag. We have to keep the deprecated code for all the deprecation period and we are not babysitters, our users are adults, they can read a deprecation message during 1 year

Reply:
> This is a user decision, if end-user decides to use a deprecated feature as default, it's totally their problem IMO, we cannot do anything there as they can enable the flag. We have to keep the deprecated code for all the deprecation period and we are not babysitters, our users are adults, they can read a deprecation message during 1 year

You are assuming they are aware of deprecations, though, which is not an assumption that I would make. Not everyone is closely watching KEDA.

When building products it's best to help customers migrate, or they will not do it until it is (almost) too late

Reply:
> You are assuming they are aware of deprecations, though, which is not an assumption that I would make. Not everyone is closely watching KEDA.

We are going to log them, if they aren't aware after a year of warnings, they will try to update KEDA, facing the "breaking change", I mean, they can be aware if they want, and they won't be aware if they don't. We are going to publish deprecation on several channels, including the pod logs.

> When building products it's best to help customers migrate, or they will not do it until it is (almost) too late

Agreed, but I prefer to keep it simple, as this is not like k8s upgrade that upgrades a lot of components, this is just rolling back to the previous version which means a few clicks

Reply:
> > I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable them (apart from the needed code, ofc). We are already giving a year for the migration
> 
> This is to get end-users off of it faster instead of only after 1 year.
> 
> > I'd reduce it too, but we decided 4 xD
> 
> Exactly, why did we decide to 4 if we change it again :)

well, we said that we will revisit this version of the proposal 😄  I am okay with keeping 4 releases, but after I gave it a thought, it seems to be quite long for me.

Reply:
Let's lave it out